### PR TITLE
feat: distributed rate limiting

### PR DESCRIPTION
**Pull request by jensneuse**

### CODE REVIEW

File: router/pkg/config/config.go [hunk 1]
```diff
@@ -215,6 +215,28 @@ type AuthorizationConfiguration struct {
...
+	Enabled        bool                    `yaml:"enabled" default:"false" envconfig:"RATE_LIMIT_ENABLED"`
+	Strategy       string                  `yaml:"strategy" default:"simple" envconfig:"RATE_LIMIT_STRATEGY" validate:"oneof=simple"`
+	SimpleStrategy RateLimitSimpleStrategy `yaml:"simple_strategy"`
+	Storage        RedisConfiguration      `yaml:"storage"`
+	// Debug ensures that retryAfter and resetAfter are set to stable values for testing
+	Debug bool `yaml:"debug" default:"false" envconfig:"RATE_LIMIT_DEBUG"`
+}
+
+type RedisConfiguration struct {
+	Addr      string `yaml:"addr" default:"localhost:6379" envconfig:"REDIS_ADDR" validate:"required"`
```
Comment (StarpTech) [inline router/pkg/config/config.go:228]: We should also support TLS configuration.

File: router/pkg/config/config.go [hunk 2]
```diff
@@ -215,6 +215,28 @@ type AuthorizationConfiguration struct {
...
+	Password  string `yaml:"password" envconfig:"REDIS_PASSWORD"`
+	KeyPrefix string `yaml:"key_prefix" default:"cosmo_rate_limit" envconfig:"RATE_LIMIT_REDIS_KEY_PREFIX" validate:"required"`
+}
+
+type RateLimitSimpleStrategy struct {
+	Rate                             int           `yaml:"rate" default:"10" envconfig:"RATE_LIMIT_SIMPLE_RATE" validate:"required,min=1"`
+	Burst                            int           `yaml:"burst" default:"10" envconfig:"RATE_LIMIT_SIMPLE_BURST" validate:"required,min=1"`
+	Period                           time.Duration `yaml:"period" default:"1s" envconfig:"RATE_LIMIT_SIMPLE_PERIOD" validate:"required,min=1s"`
+	RejectExceedingRateLimitRequests bool          `yaml:"reject_exceeding_rate_limit_requests" default:"false" envconfig:"RATE_LIMIT_SIMPLE_REJECT_EXCEEDING_RATE_LIMIT_REQUESTS"`
```
Comment (StarpTech) [inline router/pkg/config/config.go:237]: I think we can remove "RateLimit" from the name because we are inside the RateLimit section already.

**Comment by StarpTech** [comment]
Should the title not be "feat: Distributed Rate limiting"?

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router-tests/ratelimit_test.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+					},
+					Debug: true,
+				}),
+			},
+		}, func(t *testing.T, xEnv *testenv.Environment) {
+			res := xEnv.MakeGraphQLRequestOK(testenv.GraphQLRequest{
+				Query:     `query ($n:Int!) { employee(id:$n) { id details { forename surname } } }`,
+				Variables: json.RawMessage(`{"n":1}`),
+			})
+			require.Equal(t, `{"data":{"employee":{"id":1,"details":{"forename":"Jens","surname":"Neuse"}}},"extensions":{"rateLimit":{"requestRate":1,"remaining":1,"retryAfterMs":1234,"resetAfterMs":1234}}}`, res.Body)
```
[the remark this answers is not in the document]
Comment (jensneuse) [reply router-tests/ratelimit_test.go:129]: No, I will add extension codes for Authorization & Rate Limiting in one PR, but not before 2 weeks. Not enough time and this PR already addresses too many things.

File: router-tests/ratelimit_test.go [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+			res = xEnv.MakeGraphQLRequestOK(testenv.GraphQLRequest{
+				Query:     `query ($n:Int!) { employee(id:$n) { id details { forename surname } } }`,
+				Variables: json.RawMessage(`{"n":1}`),
+			})
+			require.Equal(t, `{"data":{"employee":{"id":1,"details":{"forename":"Jens","surname":"Neuse"}}},"extensions":{"rateLimit":{"requestRate":1,"remaining":0,"retryAfterMs":1234,"resetAfterMs":1234}}}`, res.Body)
```
[the remark this answers is not in the document]
Comment (jensneuse) [reply router-tests/ratelimit_test.go:134]: As per the GraphQL spec, extensions is an object.

File: router/pkg/config/config.go (hunk 1 above)
Comment (jensneuse) [reply router/pkg/config/config.go:228]: Intentionally left it out for simplicity.
Let's get feedback.

File: router/pkg/config/config.go (hunk 2 above)
Comment (jensneuse) [reply router/pkg/config/config.go:237]: Apollo uses the term "reject" for authorization, which I also adopted for authorization and thought I'd stay with the naming. This is an MVP, we can still change a lot of things based on feedback.

File: router-tests/ratelimit_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (StarpTech) [reply router-tests/ratelimit_test.go:134]: ok

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**